meter/atlas_registry.cc

```diff
@@ -0,0 +1,177 @@
+#include "atlas_registry.h"
+#include "../interpreter/group_by.h"
+#include "../interpreter/interpreter.h"
+#include "../util/logger.h"
+#include "default_counter.h"
+#include "default_distribution_summary.h"
+#include "default_gauge.h"
+#include "default_max_gauge.h"
+#include "default_long_task_timer.h"
+#include "default_timer.h"
+
+#include <numeric>
+#include <sstream>
+
+using atlas::util::Logger;
+
+namespace atlas {
+namespace meter {
+
+SystemClock system_clock;
+
+// we mostly use this class to avoid adding an external dependency to
+// the interpreter from our public headers
+class AtlasRegistry::impl {
+ public:
+  explicit impl() noexcept {}
+
+  std::shared_ptr<Meter> GetMeter(IdPtr id) noexcept {
+    std::lock_guard<std::mutex> lock(meters_mutex);
+    auto maybe_meter = meters_.find(id);
+    if (maybe_meter != meters_.end()) {
+      return maybe_meter->second;
+    }
+    return std::shared_ptr<Meter>(nullptr);
+  }
+
+  // only insert if it doesn't exist, otherwise return the existing meter
+  std::shared_ptr<Meter> InsertIfNeeded(std::shared_ptr<Meter> meter) noexcept {
+    std::lock_guard<std::mutex> lock(meters_mutex);
```

Comment:
I guess there isn't a convenient concurrent map implementation. With the java version we saw heavy contention with `computeIfAbsent` so I'm guessing the lock would also have that problem if used with many threads.